### PR TITLE
Lex dot-idents for `#if`/`#elseif` conditionals

### LINKED ISSUE #8032: Lord of the Parser - The Two Streams
I tried making `#if target.sys` parse today and ran into some trouble with our stream handling. There are two streams:

* `s`: The normal stream which goes through `process_token` and supports comments as well as conditional compilation
* `sraw`: The raw stream used for parsing the conditions of  `#if` and `#elseif`

The first problem I noticed is that doing `ignore(Stream.peek sraw)` causes parsing to fail,  which should not happen.

The other problem that could bite us at some point is that `parse_macro_cond` goes into `expr` parsing. There's a chance that this leads to `last_token` being called at some point, which does `TokenCache.get`. However, `TokenCache` is connected to `s`, not `sraw`. This would lead to the token index of `sraw` being used to look up tokens on `s`, which could fail pretty badly.

There's yet another issue which has to do with `Stream.Failure` escaping `parse_macro_cond`. This makes `#if #if` parse, while it causes errors on `#if #if #if`.

### PR DESCRIPTION
**Pull request by Simn**
After trying an failing to solve this at parser-level, I finally realized that we can simply lex `target.sys` as a single token when dealing with conditionals.

Closes #8032

### CODE REVIEW

File: src/syntax/lexer.ml [hunk 1]
```diff
@@ -277,6 +277,20 @@ let ident = [%sedlex.regexp?
 	)
 ]
 
+let sharp_ident = [%sedlex.regexp?
```
Comment (Simn) [inline src/syntax/lexer.ml:280]: @RealyUniqueName Could you double-check the regex? The idea is to match `part.part` but not `part.` or `part..part` or other weird stuff.

Comment (Aurel300) [reply src/syntax/lexer.ml:280]: `[a-z_]+(\.[a-zA-Z0-9_]+|[a-zA-Z0-9_]*)*` (I think that's what the sedlex expression is in RE language)

This allows identifiers after the first one to begin with numerals or uppercase letters, e.g. `target.9` – do you want to allow that?

Comment (Aurel300) [reply src/syntax/lexer.ml:280]: How about

`[a-z_][a-zA-Z0-9_]*(\.[a-z_][a-zA-Z0-9_]*)*`

Comment (Simn) [reply src/syntax/lexer.ml:280]: Yes good point. Could you translate that to sedlex and push it?

**Pull request merged**